native/dbdsqr.go

```diff
@@ -0,0 +1,493 @@
+		maxIt := maxIter * n * n
+		iter := 0
+		oldll := -1
+		oldm := -1
+		// m points to the last element of unconverged part of matrix.
+		m := n
+
+	Outer:
+		for {
+			if m <= 1 {
+				break
+			}
+			if iter > maxIt {
+				info = 0
+				for i := 0; i < n-1; i++ {
+					if e[i] != 0 {
+						info++
+					}
+				}
+				return info == 0
+			}
+			// Find diagonal block of matrix to work on.
+			if tol < 0 && d[m-1] <= thresh {
+				d[m-1] = 0
+			}
+			smax = math.Abs(d[m-1])
+			smin := smax
+			var ll int
+			var broke bool
+			for lll := 0; lll < m-1; lll++ {
+				ll = m - lll - 2
+				abss := math.Abs(d[ll])
+				abse := math.Abs(e[ll])
+				if tol < 0 && abss <= thresh {
+					d[ll] = 0
+				}
+				if abse <= thresh {
+					broke = true
+					break
+				}
+				smin = math.Min(smin, abss)
+				smax = math.Max(math.Max(smax, abss), abse)
+			}
+			if broke {
+				e[ll] = 0
+				if ll == m-2 {
+					// Convergence of bottom singular value, return to top.
+					m--
+					continue
+				}
+			} else {
+				ll = -1
+			}
+			ll++
+			// e[ll] through e[m-2] are nonzero, e[ll-1] is zero
+			if ll == m-2 {
+				// Handle 2×2 block separately.
+				sigmn, sigmx, sinr, cosr, sin1, cos1 := impl.Dlasv2(d[m-2], e[m-2], d[m-1])
+				d[m-2] = sigmx
+				e[m-2] = 0
+				d[m-1] = sigmn
+				if ncvt > 0 {
+					bi.Drot(ncvt, vt[(m-2)*ldvt:], 1, vt[(m-1)*ldvt:], 1, cosr, sinr)
+				}
+				if nru > 0 {
+					bi.Drot(nru, u[m-2:], ldu, u[m-1:], ldu, cos1, sin1)
+				}
+				if ncc > 0 {
+					bi.Drot(ncc, c[(m-2)*ldc:], 1, c[(m-1)*ldc:], 1, cos1, sin1)
+				}
+				m -= 2
+				continue
+			}
+			// If working on a new submatrix, chose shift direction from larger end
+			// diagonal element to smaller.
+			if ll > oldm-1 || m-1 < oldll {
+				if math.Abs(d[ll]) >= math.Abs(d[m-1]) {
+					idir = 1
+				} else {
+					idir = 2
+				}
+			}
+			// Apply convergence tests.
+			if idir == 1 {
+				// Run convergence test in forward direction.
+				// First apply standard test to bottom of matrix.
+				if math.Abs(e[m-2]) <= math.Abs(tol)*math.Abs(d[m-1]) || (tol < 0 && math.Abs(e[m-2]) <= thresh) {
+					e[m-2] = 0
+					continue
+				}
+				if tol >= 0 {
+					// If relative accuracy desired, apply convergence criterion forward.
+					mu := math.Abs(d[ll])
+					smin1 = mu
+					for lll := ll; lll < m-1; lll++ {
+						if math.Abs(e[lll]) <= tol*mu {
+							e[lll] = 0
+							continue Outer
+						}
+						mu = math.Abs(d[lll+1]) * (mu / (mu + math.Abs(e[lll])))
+						smin1 = math.Min(smin1, mu)
+					}
+				}
+			} else {
+				// Run convergence test in backward direction.
+				// First apply standard test to top of matrix.
+				if math.Abs(e[ll]) <= math.Abs(tol)*math.Abs(d[ll]) || (tol < 0 && math.Abs(e[ll]) <= thresh) {
+					e[ll] = 0
+					continue
+				}
+				if tol >= 0 {
+					// If relative accuracy desired, apply convergence criterion backward.
+					mu := math.Abs(d[m-1])
+					smin1 = mu
+					for lll := m - 2; lll >= ll; lll-- {
```

Comment:
All this `l`, `ll`, `lll` make me queasy. What do you think about `l`, `l2` and `l3`? Also horrible, but easier to see errors.